src/Npgsql/Util/TimeoutCancellationTokenSourceWrapper.cs

```diff
@@ -14,43 +16,101 @@ namespace Npgsql.Util
     class TimeoutCancellationTokenSourceWrapper : IDisposable
     {
         public TimeSpan Timeout { get; set; }
-        CancellationTokenSource _timeoutCts = new CancellationTokenSource();
+        CancellationTokenSource _cts = new CancellationTokenSource();
+        CancellationTokenRegistration _registration;
 
-        public TimeoutCancellationTokenSourceWrapper() => Timeout = System.Threading.Timeout.InfiniteTimeSpan;
+#if DEBUG
+        bool _isRunning;
+#endif
+
+        public TimeoutCancellationTokenSourceWrapper() => Timeout = InfiniteTimeSpan;
 
         public TimeoutCancellationTokenSourceWrapper(TimeSpan timeout) => Timeout = timeout;
 
         /// <summary>
         /// Set the timeout on the wrapped <see cref="CancellationTokenSource"/>
         /// and make sure that it hasn't been cancelled yet
         /// </summary>
-        public CancellationToken Start()
+        /// <param name="cancellationToken">An optional cancellation token that will be linked with the
+        /// contained <see cref="CancellationTokenSource"/></param>
+        /// <returns>The <see cref="CancellationToken"/> from the wrapped <see cref="CancellationTokenSource"/></returns>
+        public CancellationToken Start(CancellationToken cancellationToken = default)
```

Comment:
How about if the `cancellationToken` is already cancelled, we just cancel the `cts` and return its token?

Reply:
> we just cancel the cts and return its token?

Yes, probably.

We could even shortcut and directly return the cancelled token in that case.

Reply:
> We could even shortcut and directly return the cancelled token in that case.

After looking into this I can say that shortcutting isn't an option because we wouldn't have a valid state in the wrapped cancellation token.